apps/pipeline/lib/pipeline/reader/topic_reader.ex

```diff
@@ -0,0 +1,109 @@
+defmodule Pipeline.Reader.TopicReader do
+  @moduledoc """
+  Implementation of `Pipeline.Reader` for Kafka topics.
+  """
+
+  use Retry
+  @behaviour Pipeline.Reader
+
+  @type init_args() :: [
+          instance: atom(),
+          connection: atom(),
+          endpoints: Elsa.endpoints(),
+          topic: String.t(),
+          handler: module()
+        ]
+
+  @impl Pipeline.Reader
+  @spec init(init_args()) :: :ok | {:error, term()}
+  @doc """
+  Sets up infrastructure to consume messages off a Kafka topic. Ensures the topic is created.
+
+  Optional arguments:
+  * `handler_init_args` - Initial state passed to handler. Defaults to `[]`.
+  * `topic_subscriber_config` - Subscriber configuration passed to Elsa. Defaults to `[]`.
+  * `retry_count` - Times to retry topic creation. Defaults to 10.
+  * `retry_delay` - Milliseconds to initially wait before retry. Defaults to 100.
+  """
+  def init(args) do
+    config = parse_args(args)
+    consumer = consumer_spec(config)
+
+    create_topic!(config)
+
+    case DynamicSupervisor.start_child(Pipeline.DynamicSupervisor, consumer) do
+      {:ok, _} -> :ok
+      {:error, {:already_started, _}} -> :ok
+      error -> {:error, error}
+    end
+  end
+
+  @impl Pipeline.Reader
+  @spec terminate(instance: atom(), topic: String.t()) :: :ok | {:error, term()}
+  @doc """
+  Destroys topic consumer infrastructure.
+  """
+  def terminate(args) do
+    with name <- :"#{args[:instance]}-#{args[:topic]}-pipeline-supervisor",
```

Comment:
you've got a couple of different methods for grabbing elements from input args/opts going on throughout. in general you seem to be leaning on inlined options (dot accessing from a map, bracket accessing here, `Keyword.fetch` below) that will blow up if the value is missing but in this case you've got `nil` being returned.
any way we can true up the methods / expected return behavior from how these are all working?

Reply:
Good points.